content/blog/elm-at-noredink/index.md

```diff
@@ -0,0 +1,445 @@
+applications. By adapting the Rails motto "convention over configuration"
+we're confident we can scale this approach indefinitely. Let's take a deeper look!
+
+## Our Elm apps
+
+At this point, you probably won't be surprised to learn that we use a
+custom wrapper as the entrypoint of our Elm programs. Here is how it looks
+like:
+
+```elm
+main : Nri.Program.Program Model Msg
+main =
+    Nri.Program.program
+        { moduleName = "Page.Interests.Main"
+        , flagsDecoder = \_ -> decode
+        , perform = perform
+        , init = init
+        , view = view
+        , update = update
+        , subscriptions = \_ -> Sub.none
+        }
+```
+
+Let's compare it to `Browser.element` from `elm/browser`:
+
+```elm
+main : Browser.Program flags model msg
+main =
+    Browser.element
+        { init = init
+        , view = view
+        , update = update
+        , subscriptions = \_ -> Sub.none
+        }
+```
+
+As you can see, the types are slightly different and we include a couple of
+different fields in our version.
+
+What are the advantages of using a custom `Nri.Program`?
+
+- reducing boilerplate: for example, we can automatically handle decoding failures so that
+  we don't have to worry about them in each individual page.
+- handling generic behaviours: for example, we can automatically detect the user's
+  input method to make our product more accessible.
+
+The differences don't stop there. Let's look at the type of the `view` function:
+
+```elm
+view : Env -> Model -> Html Msg
+```
+
+Here we take an additional `Env` argument. This record contains information
+about the current release, the logged in user and the environment where the
+code is running. For example, if we detect that we are running in test
+mode, there is no need to run fancy UI animations, thus we can disable them.
+
+If you look at the record we pass to `Nri.Program.program` you will notice
+an additional `perform` function. Let's look at the type signature:
+
+```elm
+perform : Env -> Effect -> Cmd Msg
+```
+
+This is easier to understand if we look at the type signature of `update`:
+
+```elm
+update : Env -> Msg -> Model -> ( Model, Effect )
+```
+
+The curious change is that a standard update function in Elm returns a
+`(Model, Cmd Msg)` tuple, while ours returns `(Model, Effect)`. Why is
+that? `Cmd Msg` is an opaque type that is meant to be passed as-is to the
+Elm runtime.  But in order to test that the correct side effect has been
+emitted we need to be able to inspect it: so we define our own type and we
+call it `Effect`.  Then we can use the fantastic `elm-program-test` library
+(you can find it
+[here](https://package.elm-lang.org/packages/avh4/elm-program-test/latest))
+to test our Elm programs.  In the meanwhile, in our normal application we
+can use the `perform` function to convert the `Effect` representation of a
+side effect to its `Cmd Msg` counterpart.
+
+Similarly, we have another custom wrapper for single page applications called
```

Comment:
IMO mentioning that we have a similar wrapper for SPAs would be enough. I'm really lazy and would prefer not to read the code snippet if it doesn't introduce a new concept.